test/functional/test-shell.md

```diff
@@ -0,0 +1,187 @@
+
+```
+>>> test.nodes[0].getblockchaininfo()["blocks"]
+0
+```
+
+We now let the first node generate 101 regtest blocks, and direct the coinbase
+rewards to a wallet address owned by the mining node.
+
+```
+>>> address = test.nodes[0].getnewaddress()
+>>> test.nodes[0].generatetoaddress(101, address)
+['2b98dd0044aae6f1cca7f88a0acf366a4bfe053c7f7b00da3c0d115f03d67efb', ...
+```
+Since the two nodes are both initialized by default to establish an outbound
+connection to each other during `setup`, the second node's chain will include
+the mined blocks as soon as they propagate.
+
+```
+>>> test.nodes[1].getblockchaininfo()["blocks"]
+101
+```
+The block rewards from the first block are now spendable by the wallet of the
+first node.
+
+```
+>>> test.nodes[0].getbalance()
+Decimal('50.00000000')
+```
+
+We can also log custom events to the logger.
+
+```
+>>> test.nodes[0].log.info("Successfully mined regtest chain!")
+20XX-XX-XXTXX:XX:XX.XXXXXXX TestFramework.node0 (INFO): Successfully mined regtest chain!
+```
+
+**Note: Please also consider the functional test
+[readme](../test/functional/README.md), which provides an overview of the
+test-framework**. Modules such as
+[key.py](../test/functional/test_framework/key.py),
+[script.py](../test/functional/test_framework/script.py) and
+[messages.py](../test/functional/test_framework/messages.py) are particularly
+useful in constructing objects which can be passed to the bitcoind nodes managed
+by a running TestShell object.
+
+## 5. Shutting the TestShell down
+
+Shutting down the TestShell will safely tear down all running bitcoind instances
+and remove all temporary data and logging directories.
+
+```
+>>> test.shutdown()
+20XX-XX-XXTXX:XX:XX.XXXXXXX TestFramework (INFO): Stopping nodes
+20XX-XX-XXTXX:XX:XX.XXXXXXX TestFramework (INFO): Cleaning up /path/to/bitcoin_func_test_XXXXXXX on exit
+20XX-XX-XXTXX:XX:XX.XXXXXXX TestFramework (INFO): Tests successful
+```
+To prevent the logs from being removed after a shutdown, simply set the
+`TestShell.options.nocleanup` member to `True`.
+```
+>>> test.options.nocleanup = True
+>>> test.shutdown()
+20XX-XX-XXTXX:XX:XX.XXXXXXX TestFramework (INFO): Stopping nodes
+20XX-XX-XXTXX:XX:XX.XXXXXXX TestFramework (INFO): Not cleaning up dir /path/to/bitcoin_func_test_XXXXXXX on exit
+20XX-XX-XXTXX:XX:XX.XXXXXXX TestFramework (INFO): Tests successful
+```
+
+The following utility consolidates logs from the bitcoind nodes and the
+underlying BitcoinTestFramework:
+
+* `/path/to/bitcoin/test/functional/combine_logs.py
+  '/path/to/bitcoin_func_test_XXXXXXX'`
+
+## 6. Custom TestShell parameters
+
+The TestShell object initializes with the default settings inherited from the
+BitcoinTestFramework class. The user can override these in
+`TestShell.setup(key=value)`.
+
+**Note:** `TestShell.reset()` will reset test parameters to default values and
+can be called after the TestShell is shut down.
+
+| Test parameter key | Default Value | Description |
```

Comment:
Nice! I wonder it this table should be in test/README.md which touches on some of these parameters. Best to not maintain docs in two places if feasible.

Reply:
Hm. I was thinking about that, but not sure of the following two issues if we are two consolidate test parameter documentation.

A) Minor inconsistency in parameter naming:

- `BitcoinTestFramework` command-line args vs `BitcoinTestFramework` member:
    - `--tracerpc` vs `trace_rpc`
    - `--portseed` vs `port_seed`

`TestShell` just looks up the `setup` keywords as attributes, so we pass `TestShell.setup(trace_rpc=True)` for example. I would like to avoid a translation for these two keywords, do you agree?

B) For `BitcoinTestFramework`, some of these params are overridden in `set_test_params()` and some passed in as command-line arguments, which doesn't make sense for `TestShell`, since it's interactive.

@jonatack Let me know what you think.

Reply:
The risk with documenting defaults outside the code is that they might change and the documentation goes out of date.

I tried updating the `TestShell.__doc__` string to list all the defaults programatically but that seems like overkill.

Reply:
@jachiang in reply to your question, given the good state of the PR now and the acks, I would leave it for a possible follow-up.